### ISSUE TITLE
WhitespaceAround false negative on inline array

### ISSUE THREAD

**Issue by jsotuyod** [checkstyle/checkstyle#3203]
https://checkstyle.org/config_whitespace.html#WhitespaceAround

Using Checkstyle 6.18, with the config:

``` xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="WhitespaceAround">
            <property name="ignoreEnhancedForColon" value="false"/>
        </module>
    </module>
</module>
```

The code 

``` java
public class Foo {
    public final int[] COLORS = new int[] { Color.BLACK};
    //                                                 ^ doesn't report missing whitespace before }
}
```

doesn't report that there is missing whitespace around `RCURLY`.

``` bash
$ java -jar ~/Downloads/checkstyle-6.18/checkstyle-6.18-all.jar -c config.xml Foo.java
Starting audit...
Audit done.
```
##

**Comment by romani** [comment]
please report issue as described at http://checkstyle.sourceforge.net/report_issue.html#How_to_report_a_bug

**Comment by jsotuyod** [comment]
@romani updated

**Comment by romani** [comment]
http://checkstyle.sourceforge.net/config_whitespace.html#WhitespaceAround

RCURLY is in list of defaults. There should be violation.

**Comment by rnveach** [comment]
@romani Even though `RCURLY` is a token of the check, we are specifically ignoring it because it is part of an array initialization. https://github.com/checkstyle/checkstyle/blob/39b343a32a4f1ce18ee98a34618fda0a5404d6a0/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAroundCheck.java#L416

This has been this way for years with no real issue attached to it or a reason I could find.
There are no violations currently for open or close curlies like in `new int[]{Color.BLACK};`. Should this change too?
Since the user just wants a space before, why should it be added to `WhitespaceAround` which usually deals with both sides the token?

**Comment by romani** [comment]
@rnveach ,

here is when this code appear - https://github.com/checkstyle/checkstyle/commit/b1cfd257eb46dba42033984145134eae5880f657

**Comment by Vyom-Yadav** [comment]
@romani @rnveach I understood whats basically to be done, we need to improve the check so that there is whitespace before the `RCURLY` while initializing an array. Now the doubt I have is that how to proceed to fix it, this is the first bug I am dealing with, so what all needs to be done.

**Comment by Vyom-Yadav** [comment]
Also, do I just have to generate the error message or I have to do something so that compiler automatically fixes it.

**Comment by romani** [comment]
Message should be generate on code , human will do fix

**Comment by Vyom-Yadav** [comment]
> @romani @rnveach I am on it. I understood what's basically to be done, we need to improve the check so that there is whitespace before the `RCURLY` while initializing an array. Now the doubt I have is that how to proceed to fix it, this is the first bug I am dealing with, so what all needs to be done.

@romani 3rd last line

**Comment by Vyom-Yadav** [comment]
Also if it wouldn't be added in whitespace around as user just wants space before `RCURLY` , then where would it be added and how to configure it?

**Comment by Vyom-Yadav** [comment]
@nmancus1 can you help, you seem to work on a similar issue

**Comment by nmancus1** [comment]
Personally, for most (all?) check related issues where violations aren't what we expect (false positive/ false negative), I follow the following workflow:

1. Create test input file, that contains both code that should cause violation, and code that is acceptable.
2. Create UT in appropriate file, i.e. WhitespaceAroundCheckTest in this case. 
3. Debug checkstyle executing the test you made, tracing through the execution of the check you are working on, until you see the point of failure. 
4. Modify code to prevent failure.

Take time and study the other tests in WhitespaceAroundCheckTest, and their respective input files, to help understand Checkstyle's conventions and expectations.

**Comment by Vyom-Yadav** [comment]
@romani In my local when I ran checkstyle on 
`final int[] arr = new int[] { 345, 53 };`
                                         ^------'{' is followed by whitespace. (4:38) [NoWhitespaceAfter]
So if its shows a violation for space after `LCURLY` then why would the user want a space before `RCURLY`
There is no violation for whitespace before `RCURLY`. So instead of writing a check for missing space, we should write a check for extra space just like `LCURLY`

**Comment by Vyom-Yadav** [comment]
I think that this issue is way old and a lot of changes have been made afterwards in the code base.

**Comment by Vyom-Yadav** [comment]
@romani ping

**Comment by Vyom-Yadav** [comment]
@romani give me a yes if I should make a check for extra white space or white space before.

**Comment by Vyom-Yadav** [comment]
@nmancus1 please read the conversation and tell whether my point is valid or not and whether I should start to work on adding check for no whitespace before `RCURLY`

**Comment by Vyom-Yadav** [comment]
@rnveach please read the conversation and tell whether my point is valid or not and whether I should start to work on adding check for no whitespace before `RCURLY`

**Comment by Abhishek-kumar09** [comment]
> So if its shows a violation for space after LCURLY then why would the user want a space before RCURLY

Because this can go unnoticed: `int[] arr = { 1,2};`. It is important for code to look even from both sides.

**Comment by Abhishek-kumar09** [comment]
I checked the code with config: 
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
    <module name="WhitespaceAround">
    
    <property name="tokens"
    value="ASSIGN,DIV_ASSIGN,PLUS_ASSIGN,MINUS_ASSIGN,STAR_ASSIGN,
           MOD_ASSIGN,SR_ASSIGN,BSR_ASSIGN,SL_ASSIGN,BXOR_ASSIGN,
           BOR_ASSIGN,BAND_ASSIGN,ARRAY_INIT"/>
    </module>
    </module>
</module>
```
code: 
```
class T {
  int[] arr = {1,2};
}
```
and got: 
```
 java -jar checkstyle-8.37-all.jar -c config.xml t5.java 
Starting audit...
Audit done.
```
I mean, It is not checking left or right curlies for whitespace.

**Comment by Vyom-Yadav** [comment]
@Abhishek-kumar09 It shows an error if I use a whitespace after `LCURLY`  so that means there should be no whitespace before `RCURLY` also.
![Screenshot (150)](https://user-images.githubusercontent.com/73882557/102003080-d49d5900-3d28-11eb-954d-2fa90dcc2b9f.png)
You can see there is a check saying `{` is followed by whitespace, so the user will correct it and now it will look like `int[] arr = {1,2,3 };` but there is no check saying there is whitespace before `}` . So I am proposing to write a check for that.

**Comment by Vyom-Yadav** [comment]
![Screenshot (152)](https://user-images.githubusercontent.com/73882557/102003585-2648e200-3d2f-11eb-893e-f22924643028.png)
There is no `LCURLY` token then how is this check being used? Something in `ARRAY_INIT`?

**Comment by Abhishek-kumar09** [comment]
> but there is no check saying there is whitespace before } . So I am proposing to write a check for that.

There is the check for that: https://checkstyle.org/config_whitespace.html#NoWhitespaceBefore

But currently we are here concerned with whitespaceAround (where we want whitespace before and after token).
You have to modify this check to work for lcurly, rcurly, where we want space before and after lcurly, and a space before the rcurly.

**Comment by romani** [comment]
@Vyom-Yadav , please do not discuss in this issue any other Checks.
All Checks are completely independent and should not reply on any other Check for violation.
please focus on WhitespaceAround only , https://checkstyle.org/config_whitespace.html#WhitespaceAround

**Comment by nmancus1** [comment]
Many style guides do not recommend spaces around array initializations, such as https://google.github.io/styleguide/javaguide.html#s4.8.3.1-array-initializers.

This is probably why https://github.com/checkstyle/checkstyle/blob/39b343a32a4f1ce18ee98a34618fda0a5404d6a0/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAroundCheck.java#L416 exists.  I propose that this issue be changed to create a new property, `enforceWhitespaceAroundArrayInit`, so that this feature can be turned off.  I believe this is more clear than having users discern between two different tokens we have for the same lexeme.  If we change the default tokens for this check or change the logic otherwise, this will be a breaking change for many projects, I suspect.

On the other hand, this check already has a bunch of properties; do we need to make it more complicated?

@strkkk @pbludov what do you think?

**Comment by Vyom-Yadav** [comment]
@nmancus1 I think we should create a new property, https://eslint.org/docs/rules/array-bracket-spacing although eslint is for js, still some people prefer to have spacing in arrays, especially when using strings inside arrays.

**Comment by strkkk** [comment]
I think new property will be a good solution for such case

**Comment by Vyom-Yadav** [comment]
I'll modify the existing PR

**Comment by pbludov** [comment]
Perhaps the reason is in multidimensional arrays. Consider
```java
int[][] arr = {{ 1, 2, 3, }};
```
versus
```java
int[][] arr = { { 1, 2, 3, } };
```
@Vyom-Yadav  can you do some regression testing to prove or disprove my hypothesis?

**Comment by Vyom-Yadav** [comment]
> Perhaps the reason is in multidimensional arrays. Consider
> 
> ```java
> int[][] arr = {{ 1, 2, 3, }};
> ```
> 
> versus
> 
> ```java
> int[][] arr = { { 1, 2, 3, } };
> ```
> 
> @Vyom-Yadav can you do some regression testing to prove or disprove my hypothesis?

The reason for which thing? If you run the code you gave with following configuration.
**With default config**
```
int[][] arr = {{ 1, 2, 3, }};
```
No violation.
```
int[][] arr = { { 1, 2, 3, } };
```
Still no violation.

**With new property set to `true` and tokens `RCURLY`, `ARRAY_INIT`**
```
int[][] arr = {{ 1, 2, 3, }}; 
```
violation, whitespace before `{` and after `}` ( there will be more violations but 2 of them will be repetitive )
```
int[][] arr = { { 1, 2, 3, } };
```
No violations.

**LIMITATION**
If you run the given code with config-
tokens = `ARRAY_INIT`
```
int[][] arr = {{ 1, 2, 3, }};
```
There will be violation for whitespace between `{{` . Unfortunately It will not be possible to deal with this. This was there earlier also, but `ARRAY_INIT` is not in the list of default tokens, so it wont be checked by default.
